**<p style = "text-align: center">TRƯỜNG ĐẠI HỌC KHOA HỌC TỰ NHIÊN - ĐẠI HỌC QUỐC GIA TPHCM</p>**

# **<p style = "text-align: center">LẬP TRÌNH KHOA HỌC DỮ LIỆU</p>**
## **<p style = "text-align: center">Đồ án Cuối kì</p>**
### <p style = "text-align: center"> <b>Đề tài:</b> KHÁM PHÁ, THỐNG KÊ CỦA TOP CÁC CA KHÚC TRÊN SPOTIFY</p>
<p style = "text-align: center"> Sinh viên thực hiện </p>

| Mã số sinh viên | Họ và tên |
| --- | --- |
| 20120040 | Nguyễn Quang Gia Bảo |
| 20120127 | Trần Minh Tuấn Kiệt |
| 20120136 | Huỳnh Tuấn Nam |
| 20120158 | Trần Hoàng Anh Phi |

---

# **1. Thông tin về dữ liệu**<br>
Dữ liệu được sử dụng trong bài được lấy từ trang: <a href = "https://www.kaggle.com/datasets/yelexa/spotify200?select=final.csv"> https://www.kaggle.com/datasets/yelexa/spotify200?select=final.csv</a><br>
Dataset chứa các ca khúc nằm trong TOP 200 các ca khúc hay nhất trong 1 tuần từ 2021/02/04 ~ 2022/07/14

Theo như file "README.md":
> This directory contains codes that I used to create the dataset: https://www.kaggle.com/datasets/yelexa/spotify200<br><br>To create the dataset, I downloaded the csv files from the 'Weekly Top Songs' charts on Spotify Charts for each country from the week of 2021/02/04 ~ 2022/07/14. For the 'Global' charts, there is data from the week of 2016/12/29 ~ 2022/07/14. I used concat.py to create one combined file for each country, then used data.py to collect additional data for each song from the Spotify API.<br><br>The dataset was aggregated into an interactive dashboard, where you can explore artists and music trends around the world: https://public.tableau.com/app/profile/yejielee/viz/TheModernARExperienceWhichGlobalArtistWillYouSign/ar_dashboard?publish=yes

In [108]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns

In [132]:
spotify_df = pd.read_csv("csv/final.csv", index_col= 0)

C:\Users\Admin\AppData\Local\Temp\ipykernel_21436\2503757291.py:1: DtypeWarning: Columns (2,4,9,12,15,16,17,18,20,21,22,23,24,25,26,27,28,29,30,31,35) have mixed types. Specify dtype option on import or set low_memory=False.
  spotify_df = pd.read_csv("csv/final.csv", index_col= 0)


## Một số thông tin cơ bản của Data Frame

Xem vài dòng dữ liệu

In [110]:
spotify_df.head()

uri rank  artist_names artists_num artist_individual                              artist_id       artist_genre                                                        artist_img collab             track_name release_date album_num_tracks                                                       album_cover                               source peak_rank previous_rank weeks_on_chart  streams        week danceability energy   key mode loudness speechiness acousticness instrumentalness liveness valence    tempo  duration    country         region language pivot
0  spotify:track:2gpQi3hbcUAcEG8m2dlgfB    1  Paulo Londra         1.0      Paulo Londra  spotify:artist:3vQ0GE3mI0dAaxIMYe5g7z  argentine hip hop  https://i.scdn.co/image/ab6761610000e5ebf796a976c5597baf6f7b786c      0                 Plan A   2022-03-23              1.0  https://i.scdn.co/image/ab67616d0000b2737e1179e64539bedc938933ef                           WEA Latina         1             1              4  3003411  2022-04-14        0.583  0.834   0.0  1.0   -4.875      0.0444       0.0495              0.0   0.0658   0.557  173.935  178203.0  Argentina  South America  Spanish     0
1  spotify:track:2x8oBuYaObjqHqgGuIUZ0b    2           WOS         1.0               WOS  spotify:artist:5YCc6xS5Gpj3EkaYGdjyNK    argentine indie  https://i.scdn.co/image/ab6761610000e5eb75e1511f68e988110962dd9c      0           ARRANCARMELO   2022-04-06              1.0  https://i.scdn.co/image/ab67616d0000b273d8c9945c63f1806031dae6f0  DOGUITO Records / DALE PLAY Records         2           129              2  2512175  2022-04-14        0.654  0.354   5.0  1.0   -7.358      0.0738        0.724              0.0    0.134   0.262   81.956  183547.0  Argentina  South America  Spanish     0
2  spotify:track:2SJZdZ5DLtlRosJ2xHJJJa    3  Paulo Londra         1.0      Paulo Londra  spotify:artist:3vQ0GE3mI0dAaxIMYe5g7z  argentine hip hop  https://i.scdn.co/image/ab6761610000e5ebf796a976c5597baf6f7b786c      0                 Chance   2022-04-06              2.0  https://i.scdn.co/image/ab67616d0000b273274a28ec692ca28a73da1288                           WEA Latina         3            59              2  2408983  2022-04-14        0.721  0.463   1.0  0.0   -9.483      0.0646        0.241              0.0   0.0929   0.216  137.915  204003.0  Argentina  South America  Spanish     0
3  spotify:track:1O2pcBJGej0pmH2Y9XZMs6    5       Cris Mj         1.0           Cris Mj  spotify:artist:1Yj5Xey7kTwvZla8sqdsdE     urbano chileno  https://i.scdn.co/image/ab6761610000e5eb8f4ebcf4a5d23a2515374f89      0  Una Noche en Medellín   2022-01-21              1.0  https://i.scdn.co/image/ab67616d0000b273697ed12671078b5dee48f0ad                    Nabru Records LLC         5             5              8  2080139  2022-04-14         0.87  0.548  10.0  0.0   -5.253       0.077       0.0924         0.000046   0.0534   0.832   96.018  153750.0  Argentina  South America  Spanish     0
4  spotify:track:1TpZKxGnHp37ohJRszTSiq    6        Emilia         1.0            Emilia  spotify:artist:0AqlFI0tz2DsEoJlKSIiT9      pop argentino  https://i.scdn.co/image/ab6761610000e5ebaf96d1453d77d1d758c224bd      0          cuatro veinte   2022-03-24              1.0  https://i.scdn.co/image/ab67616d0000b273116b57ed14bc15d57eb838fa                     Sony Music Latin         6             9              3  1923270  2022-04-14        0.761  0.696   7.0  0.0   -3.817      0.0505       0.0811         0.000063    0.101   0.501   95.066  133895.0  Argentina  South America  Spanish     0

Số chiều của Data Frame

In [133]:
spotify_df.shape

(1787999, 35)

Tên của các cột

In [112]:
spotify_df.columns

Index(['uri', 'rank', 'artist_names', 'artists_num', 'artist_individual',
       'artist_id', 'artist_genre', 'artist_img', 'collab', 'track_name',
       'release_date', 'album_num_tracks', 'album_cover', 'source',
       'peak_rank', 'previous_rank', 'weeks_on_chart', 'streams', 'week',
       'danceability', 'energy', 'key', 'mode', 'loudness', 'speechiness',
       'acousticness', 'instrumentalness', 'liveness', 'valence', 'tempo',
       'duration', 'country', 'region', 'language', 'pivot'],
      dtype='object')

## Thông tin của một dòng

Mỗi một dòng dữ liệu là thông tin của một bài hát thịnh hành trong TOP 200 của một tuần bất kì.

## Thông tin của một cột

In [113]:
# pd.set_option('display.max_rows', None)
# pd.set_option('display.max_columns', None)
# pd.set_option('display.width', None)
# pd.set_option('display.expand_frame_repr', False)
pd.set_option('display.max_columns', None)
pd.set_option('display.expand_frame_repr', False)
pd.set_option('max_colwidth', None)
columns_describe_df = pd.read_csv("csv/columns.csv")
columns_describe_df

Tên cột                                                                                               Mô tả
0                                                                                                                          uri                                                                       URI của bản nhạc trên spotify
1                                                                                                                         rank                                                                    Xếp hạng của bản nhạc trong tuần
2                                                                                                                 artist names                                                                  Tên tất cả người thực hiện ca khúc
3                                                                                                                  artists num                                                                          Số người thực hiện bài hát
4                                                                                                            artist individual                                                                        Bài hát của cá nhân nghệ sĩ 
5                                                                                                                    artist id                                                                   URI của nghệ sĩ độc quyền bài hát
6                                                                                                                 artist genre                                                                      Thể loại mà nghệ sĩ trình diễn
7                                                                                                                   artist img                                                                      Đường dẫn hình ảnh của nghệ sĩ
8                                                                                                                       collab                                                0 nếu nghệ sĩ trình diễn một mình và 1 nếu ngược lại
9                                                                                                                   track name                                                                                    tên của bản nhạc
10                                                                                                                release date                                                                                ngày phát hành album
11                                                                                                            album num tracks                                                                             Số bản nhạc trong album
12                                                                                                                 album cover                                                                                       Bìa của album
13                                                                                                                      source                                                                                 Hãng thu âm bài hát
14                                                                                                                   peak rank                                                  Xếp hạng cao nhất của bài hát đã nhận trên Spotify
15                                                                                                               previous rank                                                               Xếp hạng của bài hát trong tuần trước
16                                                                                                              weeks on chart                                                                    Số tuần mà bài hát được xếp hạng
17                               

Để thuận tiện cho việc xử lí dữ liệu, nên chọn ra những cột thể hiện được đặc trưng của dữ liệu, một phần giúp giảm lượng dữ liệu cần giải quyết

Các cột cần thiết như: <br>
**Về bản nhạc**: uri, rank, artists_name, artists_num, artist_individual, collab, track_name, release_date, source, peak_rank, previous_rank, weeks_on_rank, streams, country, region, language<br>
**Về nghệ sĩ độc quyền của bản nhạc**: artist_id, artist_genre<br>

In [114]:
# cols = ["uri", "rank", 'artist_names', 'artists_num', 'artist_individual', 'collab', 'track_name', 'release_date', 'source', 'peak_rank', 'previous_rank', 'weeks_on_chart', 'streams', 'country', 'region', 'language', 'artist_id', 'artist_genre']
# spotify_df = spotify_df[cols]
spotify_df = spotify_df[spotify_df.columns[~spotify_df.columns.isin(['artist img','album cover', 'source', 'pivot'])]]

In [115]:
spotify_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1787999 entries, 0 to 1787998
Data columns (total 33 columns):
 #   Column             Dtype 
---  ------             ----- 
 0   uri                object
 1   rank               object
 2   artist_names       object
 3   artists_num        object
 4   artist_individual  object
 5   artist_id          object
 6   artist_genre       object
 7   artist_img         object
 8   collab             object
 9   track_name         object
 10  release_date       object
 11  album_num_tracks   object
 12  album_cover        object
 13  peak_rank          object
 14  previous_rank      object
 15  weeks_on_chart     object
 16  streams            object
 17  week               object
 18  danceability       object
 19  energy             object
 20  key                object
 21  mode               object
 22  loudness           object
 23  speechiness        object
 24  acousticness       object
 25  instrumentalness   object
 26  liveness      

**Xảy ra một số vấn đề về kiểu dữ liệu. Là do có nhiều kiểu dữ liệu bên trong cột.**

In [116]:
def open_oject_dtype(column):
    dtypes = set(column.apply(lambda x: type(x)))
    
    return dtypes

In [117]:
columns = spotify_df.columns
# open_oject_dtype(spotify_df["uri"])
spotify_df[columns].apply(open_oject_dtype)

uri                                                  {<class 'str'>}
rank                                  {<class 'str'>, <class 'int'>}
artist_names                                         {<class 'str'>}
artists_num                         {<class 'str'>, <class 'float'>}
artist_individual                                    {<class 'str'>}
artist_id                                            {<class 'str'>}
artist_genre                                         {<class 'str'>}
artist_img                          {<class 'str'>, <class 'float'>}
collab                                {<class 'str'>, <class 'int'>}
track_name                                           {<class 'str'>}
release_date                                         {<class 'str'>}
album_num_tracks                    {<class 'str'>, <class 'float'>}
album_cover                         {<class 'str'>, <class 'float'>}
peak_rank                             {<class 'str'>, <class 'int'>}
previous_rank                     

In [118]:
# spotify_df["rank"].unique()

# spotify_df["artists_num"].unique()

# spotify_df["collab"].unique()

# spotify_df["peak_rank"].unique()

# spotify_df["previous_rank"].unique()

# spotify_df["weeks_on_chart"].unique()

# spotify_df["streams"].unique()

# spotify_df[spotify_df["uri"] == "uri"]

Cột `rank`, `artists_num`, `collab`, `peak_rank`, `previous_rank`, `weeks_on_chart`, `streams`: một phần số với kiểu dữ liệu **int**, một phần số với kiểu dữ liệu là **str**<br>
Và có vài dòng trùng với tên cột. Và có 73 dòng như vậy.

Xem qua một vài thống kê của dữ liệu

In [119]:
spotify_df.describe()

uri     rank artist_names artists_num artist_individual                              artist_id artist_genre                                                        artist_img   collab                                                   track_name release_date album_num_tracks                                                       album_cover peak_rank previous_rank weeks_on_chart  streams        week        danceability   energy      key     mode loudness speechiness acousticness instrumentalness liveness              valence    tempo duration  country   region language
count                                1787999  1787999      1787999     1787999           1787999                                1787999      1787999                                                           1778680  1787999                                                      1787999      1787999          1787999                                                           1787982   1787999       1787999        1787999  1787999     1787999             1787604  1787604  1787604  1787604  1787604     1787604      1787604          1787604  1787604              1787604  1787604  1787604  1787999  1787999  1787999
unique                                 39203      401        19383          36             15044                                  15258         1875                                                             14425        5                                                        32673         3259              131                                                             27254       401           403            581   569044         291                1569     1857       37        7    18263        2624         5465             7734     2878                 2368    32794    43428       75       11       41
top     spotify:track:4iJyoBOLtHqaGxP12qzhQI       12    Bad Bunny         1.0         Bad Bunny  spotify:artist:4q3ewBCX7sLwd24euuV69X          pop  https://i.scdn.co/image/ab6761610000e5eb8ee9a6f54dcbd4bc95126b14        1  Enemy (with JID) - from the series Arcane League of Legends   2021-06-25              1.0  https://i.scdn.co/image/ab67616d0000b273e6f407c7f3a0ec98845e4431         1            -1              1  streams  2022-03-31  0.8240000000000001    0.825        1        1   -6.181      0.0309        0.237              0.0    0.111  0.46399999999999997    90.03   173381   Global   Europe  Spanish
freq                                    9360     5894        17473      515719             41290                                  41290       156910                                                             41290   614704                                                         9908        24732           488820                                                             15771    110446        110859          74609       73       23350               12527    10589   129913   517081     6496        8721         9314           622981    17846                 9323     6201     5181    95932   703860   503345

## **Kiểm tra xem các trường dữ liệu có vấn đề gì không**

**Dữ liệu bị thiếu, rỗng**

In [120]:
# spotify_df[spotify_df.isna().any(axis=1)]
# spotify_df[spotify_df.isnull().any(axis=1)]

spotify_df.isna().sum()

uri                     0
rank                    0
artist_names            0
artists_num             0
artist_individual       0
artist_id               0
artist_genre            0
artist_img           9319
collab                  0
track_name              0
release_date            0
album_num_tracks        0
album_cover            17
peak_rank               0
previous_rank           0
weeks_on_chart          0
streams                 0
week                    0
danceability          395
energy                395
key                   395
mode                  395
loudness              395
speechiness           395
acousticness          395
instrumentalness      395
liveness              395
valence               395
tempo                 395
duration              395
country                 0
region                  0
language                0
dtype: int64

Như vậy tất cả các cột đều không thiếu dữ liệu.

**Dữ liệu có bị trùng hay không?**

In [134]:
spotify_df.duplicated().value_counts()

False    1787927
True          72
dtype: int64

In [135]:
spotify_df[spotify_df.duplicated()]

uri  rank  artist_names  artists_num  artist_individual  artist_id  artist_genre  artist_img  collab  track_name  release_date  album_num_tracks  album_cover  source  peak_rank  previous_rank  weeks_on_chart  streams  week  danceability  energy  key  mode  loudness  speechiness  acousticness  instrumentalness  liveness  valence  tempo  duration  country  region  language  pivot
51224    uri  rank  artist_names  artists_num  artist_individual  artist_id  artist_genre  artist_img  collab  track_name  release_date  album_num_tracks  album_cover  source  peak_rank  previous_rank  weeks_on_chart  streams  week  danceability  energy  key  mode  loudness  speechiness  acousticness  instrumentalness  liveness  valence  tempo  duration  country  region  language  pivot
75474    uri  rank  artist_names  artists_num  artist_individual  artist_id  artist_genre  artist_img  collab  track_name  release_date  album_num_tracks  album_cover  source  peak_rank  previous_rank  weeks_on_chart  streams  week  danceability  energy  key  mode  loudness  speechiness  acousticness  instrumentalness  liveness  valence  tempo  duration  country  region  language  pivot
81333    uri  rank  artist_names  artists_num  artist_individual  artist_id  artist_genre  artist_img  collab  track_name  release_date  album_num_tracks  album_cover  source  peak_rank  previous_rank  weeks_on_chart  streams  week  danceability  energy  key  mode  loudness  speechiness  acousticness  instrumentalness  liveness  valence  tempo  duration  country  region  language  pivot
104818   uri  rank  artist_names  artists_num  artist_individual  artist_id  artist_genre  artist_img  collab  track_name  release_date  album_num_tracks  album_cover  source  peak_rank  previous_rank  weeks_on_chart  streams  week  danceability  energy  key  mode  loudness  speechiness  acousticness  instrumentalness  liveness  valence  tempo  duration  country  region  language  pivot
133873   uri  rank  artist_names  artists_num  artist_individual  artist_id  artist_genre  artist_img  collab  track_name  release_date  album_num_tracks  album_cover  source  peak_rank  previous_rank  weeks_on_chart  streams  week  danceability  energy  key  mode  loudness  speechiness  acousticness  instrumentalness  liveness  valence  tempo  duration  country  region  language  pivot
...      ...   ...           ...          ...                ...        ...           ...         ...     ...         ...           ...               ...          ...     ...        ...            ...             ...      ...   ...           ...     ...  ...   ...       ...          ...           ...               ...       ...      ...    ...       ...      ...     ...       ...    ...
1684645  uri  rank  artist_names  artists_num  artist_individual  artist_id  artist_genre  artist_img  collab  track_name  release_date  album_num_tracks  album_cover  source  peak_rank  previous_rank  weeks_on_chart  streams  week  danceability  energy  key  mode  loudness  speechiness  acousticness  instrumentalness  liveness  valence  tempo  duration  country  region  language  pivot
1705705  uri  rank  artist_names  artists_num  artist_individual  artist_id  artist_genre  artist_img  collab  track_name  release_date  album_num_tracks  album_cover  source  peak_rank  previous_rank  weeks_on_chart  streams  week  danceability  energy  key  mode  loudness  speechiness  acousticness  instrumentalness  liveness  valence  tempo  duration  country  region  language  pivot
1736824  uri  rank  artist_names  artists_num  artist_individual  artist_id  artist_genre  artist_img  collab  track_name  release_date  album_num_tracks  album_cover  source  peak_rank  previous_rank  weeks_on_chart  streams  week  danceability  energy  key  mode  loudness  speechiness  acousticness  instrumentalness  liveness  valence  tempo  duration  country  region  language  pivot
1758970  uri  rank  artist_names  artists_num  artist_individual  artist_id  artist_genre  artist_img  collab

Như vậy có đến 73 giá trị trùng nhau (72 + 1 giá trị gốc), và bằng với số lượng dòng có giá trị giống tên cột. Như vậy dữ liệu hầu như không có giá trị nào bị trùng cả.

Ta sẽ loại bỏ các dòng này và loại bỏ luôn dòng gốc 

In [136]:
spotify_df.drop_duplicates(keep=False, inplace=True)
spotify_df.duplicated().sum()

0

Xem lại các cột dữ liệu:

In [124]:
spotify_df.head()

uri rank  artist_names artists_num artist_individual                              artist_id       artist_genre                                                        artist_img collab             track_name release_date album_num_tracks                                                       album_cover peak_rank previous_rank weeks_on_chart  streams        week danceability energy   key mode loudness speechiness acousticness instrumentalness liveness valence    tempo  duration    country         region language
0  spotify:track:2gpQi3hbcUAcEG8m2dlgfB    1  Paulo Londra         1.0      Paulo Londra  spotify:artist:3vQ0GE3mI0dAaxIMYe5g7z  argentine hip hop  https://i.scdn.co/image/ab6761610000e5ebf796a976c5597baf6f7b786c      0                 Plan A   2022-03-23              1.0  https://i.scdn.co/image/ab67616d0000b2737e1179e64539bedc938933ef         1             1              4  3003411  2022-04-14        0.583  0.834   0.0  1.0   -4.875      0.0444       0.0495              0.0   0.0658   0.557  173.935  178203.0  Argentina  South America  Spanish
1  spotify:track:2x8oBuYaObjqHqgGuIUZ0b    2           WOS         1.0               WOS  spotify:artist:5YCc6xS5Gpj3EkaYGdjyNK    argentine indie  https://i.scdn.co/image/ab6761610000e5eb75e1511f68e988110962dd9c      0           ARRANCARMELO   2022-04-06              1.0  https://i.scdn.co/image/ab67616d0000b273d8c9945c63f1806031dae6f0         2           129              2  2512175  2022-04-14        0.654  0.354   5.0  1.0   -7.358      0.0738        0.724              0.0    0.134   0.262   81.956  183547.0  Argentina  South America  Spanish
2  spotify:track:2SJZdZ5DLtlRosJ2xHJJJa    3  Paulo Londra         1.0      Paulo Londra  spotify:artist:3vQ0GE3mI0dAaxIMYe5g7z  argentine hip hop  https://i.scdn.co/image/ab6761610000e5ebf796a976c5597baf6f7b786c      0                 Chance   2022-04-06              2.0  https://i.scdn.co/image/ab67616d0000b273274a28ec692ca28a73da1288         3            59              2  2408983  2022-04-14        0.721  0.463   1.0  0.0   -9.483      0.0646        0.241              0.0   0.0929   0.216  137.915  204003.0  Argentina  South America  Spanish
3  spotify:track:1O2pcBJGej0pmH2Y9XZMs6    5       Cris Mj         1.0           Cris Mj  spotify:artist:1Yj5Xey7kTwvZla8sqdsdE     urbano chileno  https://i.scdn.co/image/ab6761610000e5eb8f4ebcf4a5d23a2515374f89      0  Una Noche en Medellín   2022-01-21              1.0  https://i.scdn.co/image/ab67616d0000b273697ed12671078b5dee48f0ad         5             5              8  2080139  2022-04-14         0.87  0.548  10.0  0.0   -5.253       0.077       0.0924         0.000046   0.0534   0.832   96.018  153750.0  Argentina  South America  Spanish
4  spotify:track:1TpZKxGnHp37ohJRszTSiq    6        Emilia         1.0            Emilia  spotify:artist:0AqlFI0tz2DsEoJlKSIiT9      pop argentino  https://i.scdn.co/image/ab6761610000e5ebaf96d1453d77d1d758c224bd      0          cuatro veinte   2022-03-24              1.0  https://i.scdn.co/image/ab67616d0000b273116b57ed14bc15d57eb838fa         6             9              3  1923270  2022-04-14        0.761  0.696   7.0  0.0   -3.817      0.0505       0.0811         0.000063    0.101   0.501   95.066  133895.0  Argentina  South America  Spanish

> Ta thấy các cột `rank`,`artist_num`,`collab`, `peek_rank`, `previous_rank`, `weeks_on_chart`, `streams` lẽ ra phải có kiểu dữ liệu là numeric. 
> Do đó ta sẽ tiến hành chuyển các cột này về đúng kiểu dữ liệu:

In [125]:
def to_numeric(column, data_type):
    spotify_df[column] = spotify_df[column].astype(data_type)
    
# to_numeric('rank', int)
spotify_df[spotify_df['rank']=='rank']
# print(spotify_df)

Empty DataFrame
Columns: [uri, rank, artist_names, artists_num, artist_individual, artist_id, artist_genre, artist_img, collab, track_name, release_date, album_num_tracks, album_cover, peak_rank, previous_rank, weeks_on_chart, streams, week, danceability, energy, key, mode, loudness, speechiness, acousticness, instrumentalness, liveness, valence, tempo, duration, country, region, language]
Index: []

In [137]:
spotify_df['artists_num'].value_counts().head()

1.0    515719
2.0    349004
2.0    263304
1.0    149877
3.0    146750
Name: artists_num, dtype: int64

In [138]:
count = spotify_df['artists_num'].value_counts()
count[count.index!='1.0'].sum()

1272207

In [139]:
spotify_df['artists_num'] = spotify_df['artists_num'].apply(lambda x: str(x).replace('.0', '')).astype('Int64')
spotify_df['artists_num'].value_counts()

1     665596
2     612308
3     268533
4     101524
5      52290
6      33546
7      30674
9      10089
8       4504
11      3322
10      2780
14      1134
19       456
20       320
12       288
15       255
17       187
13       104
16        16
Name: artists_num, dtype: Int64

Có 2 giá trị 0 và 2 giá trị 1

In [140]:
spotify_df['collab'] = spotify_df['collab'].astype(int)
spotify_df['collab'].value_counts()

1    1122330
0     665596
Name: collab, dtype: int64

Cột `collab` có 665596 giá trị 0 (chỉ có 1 nghệ sĩ biểu diễn), tuy nhiên số lượng bài hát chỉ có 1 nghệ sĩ biểu diễn ở cột `artists_num` lại là 668376
 > Tại sao 2 giá trị này khác nhau? Ta sẽ thử tìm hiểu nguyên nhân của sự khác biệt này.

In [145]:
spotify_df[spotify_df['artists_num'] == 1].loc[1400].head()

uri                  spotify:track:3krzcyMKlRXjVkJFZqMN4t
rank                                                  115
artist_names                                 Paulo Londra
artists_num                                             1
artist_individual                            Paulo Londra
Name: 1400, dtype: object

Quan sát thấy cột `artist_names` có nhiều giá trị nhưng cột `artists_num` lại có giá trị bằng 1.

Vậy ta cần phải tính lại các giá trị ở cột `artist_num`.

In [131]:
# index_err = spotify_df[spotify_df['artists_num'] == 1][spotify_df['collab'] !=0].index
# spotify_df.loc[index_err, 'artists_num'] = spotify_df.loc[index_err, 'artist_names'].apply(len)
# spotify_df.loc[index_err, 'artists_num']